dhall-csv/src/Dhall/Csv.hs

```diff
@@ -25,42 +104,120 @@ import qualified Dhall.Import
 import qualified Dhall.Parser
 import qualified Dhall.Pretty
-import qualified Dhall.TypeCheck
+import qualified Dhall.TypeCheck                       as TypeCheck
 import qualified Dhall.Util
 import qualified System.FilePath
 
-data CompileError = Unsupported (Expr Void Void)
+{-| This is the exception type for errors that can arise when converting from
+    Dhall to CSV.
 
-instance Show CompileError where
-    show (Unsupported e) =
+    It contains information on the specific cases that might
+    fail to give a better insight.
+-}
+data CompileError
+    = Unsupported (Expr Void Void)
+    | NotAList (Expr Void Void)
+    | NotARecord (Expr Void Void)
+    | BareNone
+    deriving (Show)
+
+instance Exception CompileError where
+    displayException (Unsupported e) =
         Data.Text.unpack $
-            _ERROR <> ": Cannot translate to CSV                                             \n\
+            _ERROR <> ": Cannot translate record field to CSV                                \n\
+            \                                                                                \n\
+            \Explanation: Only the following types of record fields can be converted to CSV: \n\
+            \                                                                                \n\
+            \● ❰Bool❱                                                                        \n\
+            \● ❰Natural❱                                                                     \n\
+            \● ❰Integer❱                                                                     \n\
+            \● ❰Double❱                                                                      \n\
+            \● ❰Text❱                                                                        \n\
+            \● ❰Optional t❱ (where ❰t❱ is a valid record field type)                         \n\
+            \● Unions *                                                                      \n\
             \                                                                                \n\
-            \Explanation: Only records of primitive values can be                            \n\
-            \translated from Dhall to CSV.                                                   \n\
+            \* Unions can have empty alternatives or alternatives with valid                 \n\
+            \  record field types                                                            \n\
             \                                                                                \n\
             \The following Dhall expression could not be translated to CSV:                  \n\
             \                                                                                \n\
-            \" <> insert e
+            \" <> insert e <>                                                               "\n\
+            \because it has type:                                                            \n\
```

Comment:
I usually put a blank line after the insert to separate it from the prose and also lead into the prose with `...`, like this:

```suggestion
            \" <> insert e <>                                                               "\n\
            \… because it has type:                                                          \n\
```

Same comment for the other error messages